### ISSUE TITLE
Bundle enters in an irrecoverable state when a git dependency in the without group is upgraded.

### ISSUE THREAD

**Issue by rafaelfranca** [bundler/bundler#4066]
We are facing an issue where bundle enters in a irrecoverable state when a git dependecy that is in one of the groups listed in the `without` config is upgraded. I have a repository that reproduces the problem (The README contains how to reproduce the issue) https://github.com/rafaelfranca/bundle_test.

Here are the outputs:

```
$ git checkout error
$ bundle check
The Gemfile's dependencies are satisfied
$ bundle exec irb
git://github.com/modeset/teaspoon.git (at 42a4f61) is not yet checked out. Run `bundle install` first.
$ bundle install
git://github.com/modeset/teaspoon.git (at 42a4f61) is not yet checked out. Run `bundle install` first.
```

As you can see [in the diff](https://github.com/rafaelfranca/bundle_test/compare/error), only the sha of the gem that was changed.

For my debug session it is happening [in this line of the `index` method](https://github.com/bundler/bundler/blob/v1.10.6/lib/bundler/definition.rb#L211). `index` is being called in [`specs`](https://github.com/bundler/bundler/blob/v1.10.6/lib/bundler/definition.rb#L143) and `specs` is being called in [`specs_for`](https://github.com/bundler/bundler/blob/v1.10.6/lib/bundler/definition.rb#L185). Only `specs_for` has information about the groups that need to be loaded, but I think `index` should not try to load gems that are not in those groups.

I'm having a hard time to define what is the expected behavior, but I think either `bundle check` should fail and `bundle install` works.

cc @andremedeiros

**Comment by segiddins** [comment]
So the expected behavior here imho is that install will _download_ the locked sha, and then not use it

**Comment by rafaelfranca** [comment]
how about the check?

**Comment by segiddins** [comment]
Check _probably_ should return false? But I tried looking into this, and it's some quite hairy code that probably only @indirect fully understands :P

**Comment by rafaelfranca** [comment]
Yeah, I tried too and it was not easy to figure it out. Thank you for looking!

**Comment by lynnco** [comment]
This bug still exists (thanks @rafaelfranca for the good reproduction case) and is waiting for anyone available to work on a fix

**Comment by lynnco** [comment]
Ah nevermind! I needed to update the Bundler version in the lockfile for the test case. This issue is resolved in Bundler 1.13.1
